Make energy guns able to use burst fire.

phil235 opened the pull request:
- Energy guns, magic guns, and syringe guns can now use burst fire.
- Repathed obj/item/weapon/gun/projectile to obj/item/weapon/gun/ballistic to avoid confusion with actual projectile. Repathed obj/item/weapon/gun/energy/gun to obj/item/weapon/gun/energy/e_gun for obvious reasons.
- Modified the file organisation in the projectiles module. There's one folder for ammo box/magazine, and a different one for ammo casings. Renamed the relevant folders to "ballistic".
- Fixes syringe gun projectile spinning around and generally not facing the right direction.

phil235 committed:
Make energy guns able to use burst fire.

Bawhoppen commented:
Such a big change and not even a description?

phil235 commented:
@Bawhoppen Jesus, The PR's open for 7 minutes, and it's clearly stated WIP....

Bawhoppen commented:
Well I didn't imply I thought it was done but usually people will explain it before doing further work

PKPenguin321 commented:
We counterstrike now

Bawhoppen commented:
Energy guns are pistols why would they have burdt fire anyways

phil235 commented:
When I say "energy gun", I mean /obj/item/weapon/gun/energy , not actual e'guns per se.

Bawhoppen commented:
Oh you mean the code for burst fire was tied to automatic weapons path or something before?

phil235 commented:
No, burst fire code is at the gun level, but /gun/energy can't use it because of how chamber processing works on energy weapons.

optimumtact commented:
still chewing on this?

phil235 commented:
need to find the time to finish it. It's not stale I swear.

Razharas commented:
Whats burst fire?

phil235 committed:
Changed obj/item/weapon/gun/projectile to /gun/ballistic and the name of the folder from "projectile" to "ballistic" to avoid confusion between actually projectiles and guns.

Syringe gun, energy guns and magic guns can now use burst fire.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into BurstfireEnergy

# Conflicts:
#	_maps/map_files/generic/z2.dmm
#	code/game/objects/items/weapons/storage/briefcase.dm
#	code/game/objects/structures/crates_lockers/closets/secure/security.dm
#	code/modules/mob/living/simple_animal/hostile/hostile.dm
#	code/modules/projectiles/ammunition.dm
#	code/modules/projectiles/ammunition/ammo_casings.dm
#	code/modules/projectiles/boxes_magazines/external_mag.dm
#	code/modules/projectiles/guns/ballistic.dm
#	code/modules/projectiles/guns/ballistic/automatic.dm
#	code/modules/projectiles/guns/ballistic/bow.dm
#	code/modules/projectiles/projectile/reusable.dm
#	icons/obj/projectiles.dmi

phil235 committed:
fixing merge conflict shit

phil235 committed:
fixing map conflicts

phil235 committed:
more map conflict fix

phil235 commented on _maps/map_files/generic/z5.dmm:
```diff
@@ -1091,7 +1091,7 @@
 /area/mine/laborcamp)
 "dp" = (
-/obj/effect/decal/cleanable/cobweb2,
+/obj/effect/decal/cleanable/cobweb/cobweb2,
```
forgot to change this one in the obj damage refactor apparently.

phil235 commented on code/game/machinery/porta_turret/portable_turret.dm:
```diff
@@ -276,9 +276,9 @@
 				if(prob(70))
 					user << "<span class='notice'>You remove the turret and salvage some components.</span>"
 					if(installation)
-						var/obj/item/weapon/gun/energy/Gun = new installation(loc)
```
such a great var name, when you know there was an actual gun pathed as "/obj/item/weapon/gun/energy/gun". Jesus...

AnturK replied:
The whole turret gun installation needs rewrite.

phil235 commented on code/modules/projectiles/ammunition/ammo_casings.dm:
```diff
@@ -304,108 +304,3 @@
-
-// Caseless Ammunition
-
-/obj/item/ammo_casing/caseless
```
moved to its own file

phil235 commented on code/modules/projectiles/ammunition/caseless.dm:
```diff
@@ -0,0 +1,120 @@
+			else
+				user << "<span class='warning'>There's already something in [src].</span>"
+		else
+			user << "<span class='warning'>The safety cap prevents you from inserting [A] into [src].</span>"
```
Adding some feedback messages

phil235 commented on code/modules/projectiles/ammunition/caseless.dm:
```diff
@@ -0,0 +1,120 @@
+			if(!FD.pen)
+				if(!user.unEquip(A))
+					return
+				A.forceMove(FD)
```
forceMove() instead of loc =

phil235 commented on code/game/machinery/recharger.dm:
```diff
@@ -112,6 +112,7 @@
 			var/obj/item/weapon/gun/energy/E = charging
 			if(E.power_supply.charge < E.power_supply.maxcharge)
 				E.power_supply.give(E.power_supply.chargerate * recharge_coeff)
+				E.recharge_newshot()
```
since we no longer reload the chamber when pulling the trigger, we need to try to reload it whenever the energy gun's cell is recharged.

phil235 commented on code/modules/projectiles/box_magazine.dm:
```diff
@@ -0,0 +1,117 @@
+	var/multiple_sprites = 0
+	var/caliber
+	var/multiload = 1
+	var/start_empty = 0
```
new var to make ammo boxes or magazine start empty.

phil235 commented on code/modules/projectiles/firing.dm:
```diff
@@ -39,67 +39,24 @@
 	var/turf/curloc = get_turf(user)
 	if (!istype(targloc) || !istype(curloc) || !BB)
 		return 0
-	BB.ammo_casing = src
```
Var was removed, it was only used by foam darts and I changed their code.

phil235 commented on code/modules/projectiles/firing.dm:
```diff
@@ -39,67 +39,24 @@
-			target.bullet_act(BB, BB.def_zone)
-		qdel(BB)
-		BB = null
-		return 1
```
moved this piece of code inside projectile/proc/fire() (via the use of a second argument called "direct_target") for consistency.

phil235 commented on code/modules/projectiles/firing.dm:
```diff
@@ -39,67 +39,24 @@
 	var/dy = abs(target.y - current.y)
 	return locate(target.x + round(gaussian(0, distro) * (dy+2)/8, 1), target.y + round(gaussian(0, distro) * (dx+2)/8, 1), target.z)
-
-/obj/item/projectile/proc/preparePixelProjectile(atom/target, var/turf/targloc, mob/living/user, params, spread)
```
Moved to projectile.dm

phil235 commented on code/modules/projectiles/gun.dm:
```diff
@@ -43,7 +43,7 @@
 	var/obj/item/device/firing_pin/pin = /obj/item/device/firing_pin //standard firing pin for most guns
 
-	var/obj/item/device/flashlight/F = null
+	var/obj/item/device/flashlight/gun_light = null
```
Good fucking riddance, god damn.

phil235 commented on code/modules/projectiles/gun.dm:
```diff
@@ -192,7 +192,7 @@
 	return 0
 
-obj/item/weapon/gun/proc/newshot()
+obj/item/weapon/gun/proc/recharge_newshot()
```
name change so guns and ammo casings don't have different procs with the same name.

phil235 commented on code/modules/projectiles/gun.dm:
```diff
@@ -266,16 +267,16 @@ obj/item/weapon/gun/proc/newshot()
-		var/obj/item/device/flashlight/seclite/S = I
-		if(can_flashlight)
-			if(!F)
+	if(can_flashlight)
```
small simplification of the code here.

phil235 commented on code/modules/projectiles/guns/ballistic.dm:
```diff
@@ -1,212 +1,207 @@
+	var/spawnwithmagazine = 1
+	var/mag_type = /obj/item/ammo_box/magazine/m10mm //Removes the need for max_ammo and caliber info
+	var/obj/item/ammo_box/magazine/magazine
+	var/casing_ejector = 1 //whether the gun ejects the chambered casing
```
New var so we don't have lots of guns overriding process_chamber() just to change the value of an argument.

phil235 commented on code/modules/projectiles/guns/ballistic.dm:
```diff
@@ -1,212 +1,207 @@
+	var/obj/item/ammo_casing/AC = chambered //Find chambered round
+	if(istype(AC)) //there's a chambered round
+		if(casing_ejector)
+			AC.forceMove(get_turf(src)) //Eject casing onto ground.
```
forceMove() instead of loc =

phil235 commented on code/modules/projectiles/guns/ballistic/bow.dm:
```diff
@@ -44,50 +45,26 @@
-	force = 10
-	sharpness = IS_SHARP
-	projectile_type = /obj/item/projectile/bullet/reusable/arrow
-	caliber = "arrow"
```
These are moved to the correct files.

phil235 commented on code/modules/projectiles/guns/energy.dm:
```diff
@@ -62,36 +64,42 @@
 		update_icon()
 
-/obj/item/weapon/gun/energy/afterattack(atom/target as mob|obj|turf, mob/living/user as mob|obj, params)
-	newshot() //prepare a new shot
```
if this PR needed to be summarized in one line, this is it.

phil235 commented on code/modules/projectiles/guns/energy.dm:
```diff
@@ -62,36 +64,42 @@
 	return power_supply.charge >= shot.e_cost
 
-/obj/item/weapon/gun/energy/newshot()
+/obj/item/weapon/gun/energy/recharge_newshot(no_cyborg_drain)
```
the argument is here to prevent cyborgs from losing power when selecting a different ammo type for example (if we ever gave the HoS gun to a borg).

phil235 commented on code/modules/projectiles/guns/energy/nuclear.dm:
```diff
@@ -1,17 +1,17 @@
 	ammo_type = list(/obj/item/ammo_casing/energy/disabler, /obj/item/ammo_casing/energy/laser)
 	origin_tech = "combat=4;magnets=3"
-	modifystate = 2
+	modifystate = 1
```
don't remember why I changed this, but I had good reasons.

optimumtact replied:
:grinning:

phil235 commented on code/modules/projectiles/guns/energy/special.dm:
```diff
@@ -122,7 +122,6 @@
 	desc = "A mining tool capable of expelling concentrated plasma bursts. You could use it to cut limbs off of xenos! Or, you know, mine stuff."
 	icon_state = "plasmacutter"
 	item_state = "plasmacutter"
-	modifystate = -1
```
probably does nothing.

phil235 commented on code/modules/projectiles/guns/energy/special.dm:
```diff
@@ -142,10 +141,12 @@
 		var/obj/item/stack/sheet/S = A
 		S.use(1)
 		power_supply.give(1000)
+		recharge_newshot(1)
```
power cell was recharged: we try to reload the chamber if it was empty (because of lack of power).

phil235 commented on code/modules/projectiles/guns/grenade_launcher.dm:
```diff
@@ -36,7 +36,7 @@
 		fire_grenade(target,user)
 	else
-		usr << "<span class='danger'>The grenade launcher is empty.</span>"
+		user << "<span class='danger'>The grenade launcher is empty.</span>"
```
tiny fix. Also, this grenade launcher is not a real gun, it shouldn't be a child of gun and shouldn't be in the folder.

phil235 commented on code/modules/projectiles/guns/magic.dm:
```diff
@@ -23,7 +23,6 @@
 	righthand_file = 'icons/mob/inhands/items_righthand.dmi'
 
 /obj/item/weapon/gun/magic/afterattack(atom/target, mob/living/user, flag)
-	newshot()
```
same reason as energy gun

phil235 commented on code/modules/projectiles/guns/magic.dm:
```diff
@@ -66,14 +64,15 @@
 		return 0
 	charge_tick = 0
 	charges++
+	if(charges == 1)
```
we were empty so we try to reload the chamber with a  new shot

phil235 commented on code/modules/projectiles/guns/syringe_gun.dm:
```diff
@@ -18,30 +18,17 @@
 	..()
 	chambered = new /obj/item/ammo_casing/syringegun(src)
 
```
roughyl similar change to syringe gun, to allow burst fire.

phil235 commented on code/modules/projectiles/projectile.dm:
```diff
@@ -13,7 +13,6 @@
 	resistance_flags = LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
 	var/def_zone = ""	//Aiming at
 	var/mob/firer = null//Who shot it
-	var/obj/item/ammo_casing/ammo_casing = null
```
no longer used. Also not very useful since a lot can happen to the casing between the moment the proj is fired and the moment it hits something.

phil235 commented on code/modules/projectiles/projectile.dm:
```diff
@@ -244,19 +247,53 @@
 
-/obj/item/projectile/proc/dumbfire(var/dir)
-	current = get_ranged_target_turf(src, dir, world.maxx)
-	fire()
```
unused

phil235 commented on code/modules/projectiles/projectile/bullets.dm:
```diff
@@ -203,13 +203,8 @@
 
-//Piercing Syringe
-/obj/item/projectile/bullet/dart/syringe/piercing
-	piercing = 1
```
removed to simplify code, we now use a regular syringe dart but with piercing set to 1 (we check if the syringe item used is a piercer or not (new piercing_proj var)

phil235 commented on code/modules/reagents/reagent_containers/syringes.dm:
```diff
@@ -12,7 +12,7 @@
 	var/mode = SYRINGE_DRAW
 	var/busy = 0		// needed for delayed drawing of blood
-	var/projectile_type = /obj/item/projectile/bullet/dart/syringe
+	var/proj_piercing = 0 //does it thick clothes when shot with syringe gun
```
a word....

phil235 commented on code/modules/projectiles/projectile/reusable.dm:
```diff
@@ -36,24 +36,25 @@
 	icon_state = "foamdart_proj"
 	ammo_type = /obj/item/ammo_casing/caseless/foam_dart
 	range = 10
+	var/modified = 0
```
to remember that the projectile and casing are modifier. So we don't have to use the ammo_casing var that I'm removing.

phil235 committed:
two tiny fixes.

AnturK commented on code/modules/projectiles/ammunition/caseless.dm:
```diff
@@ -0,0 +1,120 @@
+	desc = "A caseless bullet casing."
+	firing_effect_type = null
+
+/obj/item/ammo_casing/caseless/fire_casing(atom/target, mob/living/user, params, distro, quiet)
```
Should probably have same arguments as base one ? Not sure you actually changed anything here.

AnturK reviewed (approved):
Looks good.

phil235 committed:
tiny tweak

Steelpoint commented:
Interesting feature, questionable viable on energy weapons since ballistics tend to have the advantage of a significantly larger magazine size AND the benefit of reloading on the field, whereas most energy weapons heavily emphasize accuracy.

phil235 commented:
Technically it's not a feature but a bugfix. Burstfire already exists as a feature but bugged with energy guns because of how they operated.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into BurstfireEnergy

# Conflicts:
#	code/game/objects/items/weapons/storage/uplink_kits.dm
#	code/modules/projectiles/guns/ballistic/bow.dm
#	code/modules/research/designs/weapon_designs.dm

phil235 committed:
fixing merge conflicts.
Moving the practice mini egun to the gun module.
Renamed nuclear.dm to energy_gun.dm

phil235 committed:
map conflict fixes
